django/utils/autoreload.py

```diff
@@ -233,91 +175,428 @@ def wrapper(*args, **kwargs):
+        sorted_files = sorted(watched_files, key=lambda p: p.parent)
+        for directory, group in itertools.groupby(sorted_files, key=lambda p: p.parent):
+            # These paths need to be relative to the parent directory
+            self._subscribe_dir(directory, [str(p.relative_to(directory)) for p in group])
+
+    def update_watches(self):
         try:
-            exit_code = restart_with_reloader()
-            if exit_code < 0:
-                os.kill(os.getpid(), -exit_code)
+            self._update_watches()
+        except Exception as ex:
+            # If the service is still available then want to throw the
+            # original exception
+            if self.check_server_status(ex):
+                raise
+
+    def _check_subscription(self, sub):
+        subscription = self.client.getSubscription(sub)
+        if not subscription:
+            return
+        logger.debug('Watchman subscription %s has results.', sub)
+        for result in subscription:
+            # When using watch-project it's not simple to get the relative
+            # directory without storing some specific state. Store
+            # the full path to the directory in the subscription name,
+            # prefixed by it's type (glob, files)
+            root_directory = Path(result['subscription'].split(':', 1)[1])
+            logger.debug('Found root directory %s', root_directory)
+            for file in result.get('files', []):
+                self.notify_file_changed(root_directory / file)
+
+    def request_processed(self, **kwargs):
+        logger.debug('Request processed. Setting update_watches event.')
+        self.processed_request.set()
+
+    def tick(self):
+        request_finished.connect(self.request_processed)
+        self.update_watches()
+
+        while True:
+            if self.processed_request.is_set():
+                self.update_watches()
+                self.processed_request.clear()
+            try:
+                self.client.receive()
+            except pywatchman.WatchmanError as ex:
+                self.check_server_status(ex)
             else:
-                sys.exit(exit_code)
-        except KeyboardInterrupt:
-            pass
+                for sub in list(self.client.subs.keys()):
+                    self._check_subscription(sub)
+            yield
 
+    def stop(self):
+        self.client.close()
+        super().stop()
+
+    def check_server_status(self, inner_ex=None):
+        """Return True if the server is available."""
+        try:
+            self.client.query("version")
+        except Exception:
+            raise WatchmanUnavailable(str(inner_ex)) from inner_ex
+        return True
+
+    @classmethod
+    def is_available(cls):
+        if not pywatchman:
+            return False
+        client = pywatchman.client(timeout=0.01)
+        try:
+            result = client.capabilityCheck()
+        except Exception:
+            # The service is down?
```

Comment:
I think this should error somehow get printed or a warning given or something (maybe only with verbose?). (in my case it's happening if pywatchman is installed, but not watchman.): `pywatchman.WatchmanError: "watchman" executable not in PATH (%s), while executing [Errno 2] No such file or directory: 'watchman': 'watchman'`

Reply:
Hmm, I like this idea. The only issue is this will be called on every reload, so it could end up spamming the console?

Can you think of a way around this? I could modify it a bit to only throw a warning in the very first invocation (which will be the main executable), but it seems hacky.

Reply:
I've implemented this. We now have a `check_availability` method that raises an exception with a message, and this is displayed to the user 👍

Reply:
Thanks!